rest/oidc_api_test.go

```diff
@@ -1,139 +1,714 @@
+			redirectURL, err = url.Parse(redirectURI)
+			require.NoError(t, err, "error parsing redirect_uri URL")
+			assert.Equal(t, tc.wantProvider, redirectURL.Query().Get(auth.OIDCAuthProvider))
+		})
+	}
+}
+
+// Returns a new OIDCProvider.
+func mockProvider(name string) *auth.OIDCProvider {
+	return &auth.OIDCProvider{
+		Name:          name,
+		ClientID:      base.StringPointer("baz"),
+		UserPrefix:    name,
+		ValidationKey: base.StringPointer("qux"),
+	}
+}
+
+// Returns an auto registration enabled provider.
+func mockProviderWithRegister(name string) *auth.OIDCProvider {
+	return &auth.OIDCProvider{
+		Name:          name,
+		ClientID:      base.StringPointer("baz"),
+		UserPrefix:    name,
+		ValidationKey: base.StringPointer("qux"),
+		Register:      true,
+	}
+}
+
+// Returns a new OIDCProvider with Register and IncludeAccessToken flags enabled.
+func mockProviderWithRegisterWithAccessToken(name string) *auth.OIDCProvider {
+	return &auth.OIDCProvider{
+		Name:               name,
+		ClientID:           base.StringPointer("baz"),
+		UserPrefix:         name,
+		ValidationKey:      base.StringPointer("qux"),
+		Register:           true,
+		IncludeAccessToken: true,
+	}
+}
+
+// Checks End to end OpenID Connect Authorization Code flow.
+func TestOpenIDConnectAuth(t *testing.T) {
+	type test struct {
+		name              string
+		providers         auth.OIDCProviderMap
+		defaultProvider   string
+		authURL           string
+		forceAuthError    forceError
+		forceRefreshError forceError
+	}
+	tests := []test{
+		{
+			// Successful new user authentication when auto registration is enabled and a single provider
+			// is configured through providers configuration. Explicitly specified to include access token in
+			// token response.
+			name: "successful user registration against single provider",
+			providers: auth.OIDCProviderMap{
+				"foo": mockProviderWithRegisterWithAccessToken("foo"),
+			},
+			defaultProvider: "foo",
+			authURL:         "/db/_oidc?provider=foo&offline=true",
+		}, {
+			// Unsuccessful new user authentication when auto registration is enabled and a single provider
+			// is configured through providers configuration. Explicitly specified to include access token in
+			// token response.
+			name: "unsuccessful user registration against single provider",
+			providers: auth.OIDCProviderMap{
+				"foo": mockProvider("foo"),
+			},
+			defaultProvider: "foo",
+			authURL:         "/db/_oidc?provider=foo&offline=true",
+			forceAuthError: forceError{
+				errorType:            noAutoRegistrationErr,
+				expectedErrorCode:    http.StatusUnauthorized,
+				expectedErrorMessage: "Invalid login",
+			},
+		}, {
+			// Make sure user registration in successful when IncludeAccessToken is false in
+			// providers configuration.
+			name: "successful user registration against single provider without access token",
+			providers: auth.OIDCProviderMap{
+				"foo": mockProviderWithRegister("foo"),
```

Comment:
Do we have any tests that cover successful authentication without auto-register (i.e. the user already exists)?  if not, do you think that needs coverage?

Reply:
There was no test that covers successful authentication without auto-register, but yes, it require coverage and included additional tests in the latest revision.